### PR TITLE
Refactor and fix reconnect and ping

### LINKED ISSUE #858: ping/pong timeout recursion
Discovered (thankfully in dev env) issue with ping/pong timeout when running with `run_forever`. Chain traceback, with multiple (10,000+) ping/pong timeout errors. At the end, our logger raised recursion error as it couldn't keep up. We are running with proxy and these parameters:
``` python
ping_timeout: int = 4
ping_interval: int = 5
``` 

``` log
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 354, in setSock
    dispatcher.read(self.sock.sock, read, check)
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 56, in read
    check_callback()
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 397, in check
    raise WebSocketTimeoutException("ping/pong timed out")
websocket._exceptions.WebSocketTimeoutException: ping/pong timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 354, in setSock
    dispatcher.read(self.sock.sock, read, check)
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 56, in read
    check_callback()
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 397, in check
    raise WebSocketTimeoutException("ping/pong timed out")
websocket._exceptions.WebSocketTimeoutException: ping/pong timed out

During handling of the above exception, another exception occurred:

....
```

### PR DESCRIPTION
**Pull request by j-a-n**
This PR fixes several reconnect issues.
Tested with and without rel.

- Dispatcher: do not reconnect in loop
- Avoid reconnect in recursion and hit maximum recursion depth
- Close websocket before reconnect to avoid unclosed ssl.SSLSocket warnings
- Start ping thread on socket connect
- Stop ping thread on socket disconnect
- Add reconnect test
- Add callback exception test

Fixes #858

### CODE REVIEW

**Comment by engn33r** [comment]
First, thank you for the PR! Especially for improving the tests. :+1:

I used the example client/server code from #858. When I repeated the steps I did before to recreate the issue (omitting step 4 of restarting the server quickly), the client behaves as expert, which is good. But when I do introduce step 4 and restart the server quickly, I get a different loop than reported in the original issue with this output:

```
ERROR:websocket:[Errno 111] Connection refused - reconnecting
Connection opened
WARNING:websocket:websocket connected
ERROR:websocket:ping/pong timed out - reconnecting
Connection opened
WARNING:websocket:websocket connected
ERROR:websocket:ping/pong timed out - reconnecting
Connection opened
WARNING:websocket:websocket connected
ERROR:websocket:ping/pong timed out - reconnecting
Connection opened
WARNING:websocket:websocket connected
ERROR:websocket:ping/pong timed out - reconnecting
Connection opened
WARNING:websocket:websocket connected
ERROR:websocket:ping/pong timed out - reconnecting
```

The question I would ask here is what is the expected behavior? My guess is if a user uses `run_forever()` but the server goes down and does not return a ping, an error should be triggered before the client continues attempting to reconnect. Once a reconnection is established, any history related to the previous connection should be forgotten. I'm not an expert with threading and haven't examined the new code close enough, but my guess here is that after a reconnect there is an artifact from the previous connection that interferes with the pong/pong timing.

In short, we should find a way to prevent the ping/pong error from looping after a reconnection.

**Comment by j-a-n** [comment]
Could be a timing issue when the server is back up very quickly.
I will try to reproduce the issue with the example client/server code.

**Comment by j-a-n** [comment]
I cannot reproduce the problem with the following code.
No matter how many times or how fast I restart the server, the client always reconnects without ending up in a loop.
What is the difference to your client code?

**server**
```
#!/usr/bin/env python
import asyncio
import websockets

async def hello(websocket, path):
    name = await websocket.recv()
    print("< {}".format(name))

    greeting = "Hello {}!".format(name)
    await websocket.send(greeting)
    print("> {}".format(greeting))

start_server = websockets.serve(hello, 'localhost', 8765)
print("Running in localhost:8765")

asyncio.get_event_loop().run_until_complete(start_server)
asyncio.get_event_loop().run_forever()
```

**client**
```
import websocket
import logging
logging.basicConfig()

websocket_url = "ws://localhost:8765"
ping_timeout = 1
ping_interval = 2

def on_message(ws, data):
    print(data)

def on_error(ws, error):
    logging.exception("error")

def on_open(ws):
    print("Connection opened")

ws = websocket.WebSocketApp(
    websocket_url,
    on_message=on_message,
    on_error=on_error,
    on_open=on_open
)

ws.run_forever(
    ping_timeout=ping_timeout,
    ping_interval=ping_interval,
	reconnect=1
)
```

**Comment by j-a-n** [comment]
Fixes #867

**Comment by engn33r** [comment]
I intend to run some more comprehensive tests on this PR next month but I don't have time to do the testing I want before then. I will plan to better outline how to reproduce the issue I had before and try some other test configurations, ideally improving the unit tests in CI in the process.

**Comment by engn33r** [comment]
Thank you for the detailed PR. The ping/pong logic was not implemented in a very elegant manner before and this improves it. One remaining TODO which will go in my backlog (unless you'd like to add it) is checking the existing documentation about ping/pong to determine if this PR changes any of it.

I finally got around to fixing the github CI tests and testing the client/server code with and without this change. As far as I can tell, this fixes at least a couple of recent issues without noticeable side effects. I will merge this PR and then release v1.5.0 with this upgrade.
